packages/@aws-cdk/aws-ecr/lib/repository.ts

```diff
@@ -0,0 +1,138 @@
+import cdk = require('@aws-cdk/cdk');
+import { cloudformation, RepositoryArn, RepositoryName } from './ecr.generated';
+import { Action, CountType, CountUnit, LifecycleRule, TagStatus } from './lifecycle';
+import { RepositoryRef } from "./repository-ref";
+
+export interface RepositoryProps {
+    /**
+     * Name for this repository
+     *
+     * @default Automatically generated name.
+     */
+    repositoryName?: string;
+
+    /**
+     * Life cycle rules to apply to this registry
+     *
+     * @default No life cycle rules
+     */
+    lifecycleRules?: LifecycleRule[];
+
+    /**
+     * The AWS account ID associated with the registry that contains the repository.
+     *
+     * @default The default registry is assumed.
+     */
+    lifecycleRegistryId?: string;
+
+    /**
+     * Retain the registry on stack deletion
+     *
+     * If you don't set this to true, the registry must be empty, otherwise
+     * your stack deletion will fail.
+     *
+     * @default false
+     */
+    retain?: boolean;
+}
+
+/**
+ * Define an ECR repository
+ */
+export class Repository extends RepositoryRef {
+    public readonly repositoryName: RepositoryName;
+    public readonly repositoryArn: RepositoryArn;
+    private readonly lifecycleRules = new Array<LifecycleRule>();
+    private readonly registryId?: string;
+    private policyDocument?: cdk.PolicyDocument;
+
+    constructor(parent: cdk.Construct, id: string, props: RepositoryProps = {}) {
+        super(parent, id);
+
+        const resource = new cloudformation.RepositoryResource(this, 'Resource', {
+            repositoryName: props.repositoryName,
+            // It says "Text", but they actually mean "Object".
+            repositoryPolicyText: this.policyDocument,
+            lifecyclePolicy: new cdk.Token(() => cdk.resolve(this.renderLifecyclePolicy())),
+        });
+
+        if (props.retain) {
+            resource.options.deletionPolicy = cdk.DeletionPolicy.Retain;
+        }
+
+        this.registryId = props.lifecycleRegistryId;
+        if (props.lifecycleRules) {
+            props.lifecycleRules.forEach(this.addLifecycleRule.bind(this));
+        }
+
+        this.repositoryName = resource.ref;
+        this.repositoryArn = resource.repositoryArn;
+    }
+
+    public addToResourcePolicy(statement: cdk.PolicyStatement) {
+        if (this.policyDocument === undefined) {
+            this.policyDocument = new cdk.PolicyDocument();
+        }
+        this.policyDocument.addStatement(statement);
+    }
+
+    public addLifecycleRule(rule: LifecycleRule) {
+        this.lifecycleRules.push(rule);
+    }
+
+    private renderLifecyclePolicy(): cloudformation.RepositoryResource.LifecyclePolicyProperty | undefined {
+        let lifecyclePolicyText: any;
+
+        if (this.lifecycleRules.length === 0 && !this.registryId) { return undefined; }
+
+        if (this.lifecycleRules.length > 0) {
```

Comment:
What does it means to specify a registry ID without rules?

Reply:
I don't know, but I don't want to preclude it.